### PR TITLE
Merging #20557 (Asynchronous inserts)

### PR DESCRIPTION
**Pull request by CurtizJ**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Merging #20557.

TODO:

- [ ] Better handling of huge number of async inserts with wait. Now such queries will fail, when `max_concurrent_queries_for_all_users` is reached. (Probably not in this PR).

### CODE REVIEW

**Comment by CurtizJ** [comment]
Flaky tests failure most likely related to known problem: deadlock in `TRUNCATE` query in non-replicated `MergeTree`. `TRUNCATE` queries as well as some other queries are run in `02015_async_inserts_stress_long`.

```
<Error> executeQuery: Code: 473. DB::Exception: WRITE locking attempt on "test_d0l8d0.async_inserts" has timed out! (120000ms) Possible deadlock avoided. Client should retry. (DEADLOCK_AVOIDED) (version 21.11.1.8088) (from [::1]:37514) (comment: '/usr/share/clickhouse-test/queries/0_stateless/02015_async_inserts_stress_long.sh') (in query: TRUNCATE TABLE async_inserts)
```

File: src/Core/Settings.h [hunk 1]
```diff
@@ -601,6 +601,15 @@ class IColumn;
     M(Bool, output_format_write_statistics, true, "Write statistics about read rows, bytes, time elapsed in suitable output formats.", 0) \
     M(Bool, output_format_pretty_row_numbers, false, "Add row numbers before each row for pretty output format", 0) \
     M(Bool, insert_distributed_one_random_shard, false, "If setting is enabled, inserting into distributed table will choose a random shard to write when there is no sharding key", 0) \
+    \
+    M(UInt64, async_insert_threads, 16, "Maximum number of threads to actually parse and insert data in background. Zero means asynchronous mode is disabled", 0) \
```
Comment (alexey-milovidov) [inline src/Core/Settings.h:605]: You've added it to the section of FormatFactorySettions, better to move.

File: src/Core/Settings.h [hunk 2]
```diff
@@ -601,6 +601,15 @@ class IColumn;
...
+    M(Bool, async_insert_mode, false, "Insert query is processed almost instantly, but an actual data queued for later asynchronous insertion", 0) \
+    M(Bool, wait_for_async_insert, true, "If true wait for processing of asynchronous insertion", 0) \
+    M(Seconds, wait_for_async_insert_timeout, DBMS_DEFAULT_LOCK_ACQUIRE_TIMEOUT_SEC, "Timeout for waiting for processing asynchronous insertion", 0) \
+    M(UInt64, async_insert_max_data_size, 1000000, "Maximum size in bytes of unparsed data collected per query before being inserted", 0) \
+    M(Milliseconds, async_insert_busy_timeout, 200, "Maximum time to wait before dumping collected data per query since the first data appeared", 0) \
```
Comment (alexey-milovidov) [inline src/Core/Settings.h:610]: Should have `ms` suffix as other similar settings.

File: src/Core/Settings.h [hunk 3]
```diff
@@ -601,6 +601,15 @@ class IColumn;
     M(Bool, output_format_write_statistics, true, "Write statistics about read rows, bytes, time elapsed in suitable output formats.", 0) \
     M(Bool, output_format_pretty_row_numbers, false, "Add row numbers before each row for pretty output format", 0) \
     M(Bool, insert_distributed_one_random_shard, false, "If setting is enabled, inserting into distributed table will choose a random shard to write when there is no sharding key", 0) \
+    \
+    M(UInt64, async_insert_threads, 16, "Maximum number of threads to actually parse and insert data in background. Zero means asynchronous mode is disabled", 0) \
+    M(Bool, async_insert_mode, false, "Insert query is processed almost instantly, but an actual data queued for later asynchronous insertion", 0) \
+    M(Bool, wait_for_async_insert, true, "If true wait for processing of asynchronous insertion", 0) \
+    M(Seconds, wait_for_async_insert_timeout, DBMS_DEFAULT_LOCK_ACQUIRE_TIMEOUT_SEC, "Timeout for waiting for processing asynchronous insertion", 0) \
+    M(UInt64, async_insert_max_data_size, 1000000, "Maximum size in bytes of unparsed data collected per query before being inserted", 0) \
```
Comment (alexey-milovidov) [inline src/Core/Settings.h:609]: Too large for default, I'd better set it to 100 000.

File: src/Core/Settings.h [hunk 4]
```diff
@@ -601,6 +601,15 @@ class IColumn;
     M(Bool, output_format_write_statistics, true, "Write statistics about read rows, bytes, time elapsed in suitable output formats.", 0) \
     M(Bool, output_format_pretty_row_numbers, false, "Add row numbers before each row for pretty output format", 0) \
     M(Bool, insert_distributed_one_random_shard, false, "If setting is enabled, inserting into distributed table will choose a random shard to write when there is no sharding key", 0) \
+    \
+    M(UInt64, async_insert_threads, 16, "Maximum number of threads to actually parse and insert data in background. Zero means asynchronous mode is disabled", 0) \
+    M(Bool, async_insert_mode, false, "Insert query is processed almost instantly, but an actual data queued for later asynchronous insertion", 0) \
```
Comment (alexey-milovidov) [inline src/Core/Settings.h:606]: The description is not correct. With `wait_for_async_insert` it means that the data is grouped with other similar INSERTs and we wait for actual insertion before returning result to user.

Comment (alexey-milovidov) [reply src/Core/Settings.h:606]: Why it's named `async_insert_mode`, maybe just `async_insert`?

**Pull request merged**